Add support for Texas Instruments ELF .out files

dbouras opened the pull request:
Add support for ELF .out files from the Texas Instruments (TI) linker (lnk2000 -- https://www.ti.com/lit/ug/spru513z/spru513z.pdf); readelf now matches output from the TI Object File Display utility (ofd2000); tested under Linux with Code Composer Studio version 12.8.1 (https://www.ti.com/tool/download/CCSTUDIO/12.8.1)

sevaa commented:
Did you test against the readelf that ships with pyelftools, or the one on your system?

dbouras commented:
Hi @sevaa,
Testing aside, I guess I should have added an example ELF file to illustrate the benefit of my contribution -- sorry about that :)

I will create a generic .out file (can't contribute the proprietary one I used for testing) and follow-up here with example output from readelf.py (with and without the --TI option that I added) and TI's ofd2000.

dbouras commented:
@sevaa, I stand corrected and I see your point: it shouldn't fail any of the existing tests; obviously I've missed something that needs to be "gated" on the "--TI" option. I will fix and update the PR.
Many thanks for the quick feedback!

dbouras commented:
All done; tests pass without issues now. Will add a TI ELF file and corresponding test later tonight.
Many thanks once again @sevaa!

sevaa commented:
**Did you test your TI binaries against the version of `readelf` that ships with pyelftools?** I'm finding it hard to believe that something as basic as header format in the section dump is conditional upon TI/not TI in the GNU `readelf`.

dbouras commented:
Thank you both for your comments and apologies for not providing more context; I'll try to do so below. As promised, I have generated a TI binary for the TMS320C2000 (from [this example project](https://dev.ti.com/tirex/explore/node?node=A__AMJRCjyqPWotKFi5OZWM7A__digital_power_c2000ware_sdk_software_package__5C6SEVO__LATEST&search=BUCK-DC-DC), if anyone is interested in the detail) and added it as `test/testfiles_for_readelf/ti_c2000.elf`. Running `test/all_tests.py` fails (as expected).

My motivation (when I ran into `pyelftools` a couple of weeks ago) was to understand why the output from GNU readelf and the proprietary [ofd2000 tool from TI](https://www.ti.com/lit/ug/spru513z/spru513z.pdf) had notable differences when processing a binary generated from a project I'm working on. And to use this understanding to build another internal tool of ours using `elftools`.

So let's start with the header:
readelf:
```
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF32
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           Texas Instruments TMS320C2000 DSP family
  Version:                           0x1
  Entry point address:               0x80000
  Start of program headers:          144568 (bytes into file)
  Start of section headers:          144888 (bytes into file)
  Flags:                             0x0
  Size of this header:               52 (bytes)
  Size of program headers:           32 (bytes)
  Number of program headers:         10
  Size of section headers:           40 (bytes)
  Number of section headers:         38
  Section header string table index: 37
```
versus readelf.py:
```
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00
  Class:                             ELF32
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           <unknown>
  Version:                           0x1
  Entry point address:               0x80000
  Start of program headers:          144568 (bytes into file)
  Start of section headers:          144888 (bytes into file)
  Flags:                             0x0
  Size of this header:               52 (bytes)
  Size of program headers:           32 (bytes)
  Number of program headers:         10
  Size of section headers:           40 (bytes)
  Number of section headers:         38
  Section header string table index: 37
```
The Machine field is `<uknown>`; very straightforward change here (I also added MIPS RS3 but that is irrelevant to this discussion):
```
diff --git a/elftools/elf/descriptions.py b/elftools/elf/descriptions.py
index 94b3b25..72af45a 100644
--- a/elftools/elf/descriptions.py
+++ b/elftools/elf/descriptions.py
@@ -400,6 +400,7 @@ _DESCR_E_MACHINE = dict(
     EM_MIPS='MIPS R3000',
     EM_S370='IBM System/370',
     EM_S390='IBM S/390',
+    EM_MIPS_RS3_BE='MIPS RS3000 little-endian',
     EM_MIPS_RS4_BE='MIPS 4000 big-endian',
     EM_IA_64='Intel IA-64',
     EM_X86_64='Advanced Micro Devices X86-64',
@@ -410,6 +411,10 @@ _DESCR_E_MACHINE = dict(
     EM_PPC='PowerPC',
     EM_PPC64='PowerPC64',
     EM_RISCV='RISC-V',
+    EM_MSP430='Texas Instruments embedded microcontroller MSP430',
+    EM_TI_C6000='Texas Instruments TMS320C6000 DSP Family',
+    EM_TI_C2000='Texas Instruments TMS320C2000 DSP Family',
+    EM_TI_C5500='Texas Instruments TMS320C55x DSP Family',
     EM_LOONGARCH='LoongArch',
     RESERVED='RESERVED',
 )
```

More importantly now, let's move to the sections. One of the things I wanted to do was understand how to calculate the physical LOAD address, given the RUN address read from the section header and the segment physical address, both provided by `readelf` or `readelf.py`. This normally should be:
```
load_addr = segment['p_paddr'] + (section['sh_offset'] - segment['p_offset'])
```
While trying to reconcile with the output from TI's `ofd2000` I found out the following (I illustrate the logic by copying code from my changes in readelf.py):
```
    # compute section LOAD address from the RUN address (section['sh_addr'])
    load_addr = section['sh_addr']
    for segment in self.elffile.iter_segments():
        if segment.section_in_segment(section):
            if segment['p_paddr'] == segment['p_vaddr']:
                load_addr = section['sh_addr']
            else:
                if do_TI_addr:
                    # TI ELF files need this instead:
                    load_addr = segment['p_paddr'] + ((section['sh_offset'] - segment['p_offset']) >> 1)
                    # because addresses point to 16-bit words in memory not 8-bit bytes
                else:
                    load_addr = segment['p_paddr'] + (section['sh_offset'] - segment['p_offset'])
                break

    flags = section['sh_flags']
    TI = do_TI_addr
    if (flags & SH_FLAGS.SHF_ALLOC) == 0:
        TI = False
```

Two important things to note here:
1. `section['sh_size']` for TI generated ELF files is interpreted (and printed out) differently by `ofd2000` as compared to `readelf`/`readelf.py` because (apparently) TI "defines" bytes as holding 16 bits in their DSP processors
2.  the `TI` flag (generated at the end of the code fragment above) is used for interpreting `section['sh_size']`: for sections allocated in memory it's transformed to words (dividing by 2); for sections with `aloc: N`, it's left as is.

To make a long story short, I understand that all my contribution may be of limited value if the purpose here is to just match what GNU `readelf` is doing, but I thought I could augment `readelf.py` to treat TI ELF files in the same way as TI's proprietary `ofd2000` tool.

Thank you for reading this far -- I will close with an illustrative example. This is the section output from `readelf.py`:
```
% python3.11 scripts/readelf.py -S test/testfiles_for_readelf/ti_c2000.elf 
There are 38 section headers, starting at offset 0x235f8

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] codestart         PROGBITS        00080000 000034 000004 00  AX  0   0  4
  [ 2] .cinit            PROGBITS        00080004 000038 000074 00   A  0   0  4
  [ 3] .init_array       NOBITS          00000000 003d30 000000 00  WA  0   0  4
  [ 4] .switch           NOBITS          00000000 000000 000000 00      0   0  4
  [ 5] .const            NOBITS          00000000 000000 000000 00      0   0  4
  [ 6] .reset            NOBITS          003fffc0 000000 000000 00      0   0  1
  [ 7] .stack            NOBITS          00000400 003d2c 000800 00  WA  0   0  2
  [ 8] .data             NOBITS          00008964 003d2c 000094 00  WA  0   0  2
  [ 9] .bss              NOBITS          00008000 003d2c 0012c6 00  WA  0   0  2
  [10] .sysmem           NOBITS          00000000 000000 000000 00      0   0  1
  [11] .TI.ramfunc       PROGBITS        0000c000 002e4c 00096a 00 WAX  0   0  2
  [12] ramfuncs          PROGBITS        0000c4b5 0037b8 00013c 00  AX  0   0  1
  [13] SFRA_F32_Data     NOBITS          00000000 000000 000000 00      0   0 64
  [14] FPUmathTables     PROGBITS        0008b000 0038f4 0001a0 00   A  0   0  2
  [15] .scratchpad       NOBITS          00000000 000000 000000 00      0   0  1
  [16] .bss_cla          NOBITS          00000000 003d30 000000 00  WA  0   0  1
  [17] controlVariables  NOBITS          00000000 000000 000000 00      0   0  1
  [18] .const_cla        NOBITS          00009000 003d30 000000 00   A  0   0  1
  [19] isrcodefuncs      PROGBITS        00009000 003a94 000296 00  AX  0   0  1
  [20] dclfuncs          NOBITS          0000914b 003d30 000000 00   A  0   0  1
  [21] Cla1Prog          NOBITS          00009000 003d30 000000 00   A  0   0  4
  [22] .ppdata           NOBITS          00000000 000000 000000 00      0   0  1
  [23] .debug_types      PROGBITS        00000000 003d30 00a4a0 00      0   0  1
  [24] .debug_info       PROGBITS        00000000 00e1d0 003db2 00      0   0  1
  [25] .debug_line       PROGBITS        00000000 011f82 005cd6 00      0   0  1
  [26] .debug_frame      PROGBITS        00000000 017c58 000db0 00      0   0  4
  [27] .text.1           PROGBITS        00081000 0000ac 002000 00  AX  0   0  1
  [28] .text.2           PROGBITS        00082000 0020ac 000d9e 00  AX  0   0  1
  [29] .debug_abbrev     PROGBITS        00000000 018a08 00048a 00      0   0  0
  [30] .debug_str        PROGBITS        00000000 018e92 001cf3 00      0   0  0
  [31] .debug_aranges    PROGBITS        00000000 01ab85 000b38 00      0   0  0
  [32] .debug_pubnames   PROGBITS        00000000 01b6bd 0019fd 00      0   0  0
  [33] __TI_build_attrib <unknown>       00000000 01d0ba 00003a 00      0   0  0
  [34] .symtab           SYMTAB          00000000 01d0f4 004040 10     36 735  0
  [35] .TI.section.flags <unknown>       00000000 021134 000037 00      0   0  0
  [36] .strtab           STRTAB          00000000 02116b 0021b7 01   S  0   0  0
  [37] .shstrtab         STRTAB          00000000 023322 000193 01   S  0   0  0
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  p (processor specific)
```
and the segment output for reference:
```
% python3.11 scripts/readelf.py -l test/testfiles_for_readelf/ti_c2000.elf 

Elf file type is EXEC (Executable file)
Entry point is 0x80000
There are 10 program headers, starting at offset 144568

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000034 0x00080000 0x00080000 0x00004 0x00004 R E 0x4
  LOAD           0x000038 0x00080004 0x00080004 0x00074 0x00074 R   0x4
  LOAD           0x0000ac 0x00081000 0x00081000 0x02d9e 0x02d9e R E 0x1
  LOAD           0x002e4c 0x0000c000 0x00086000 0x0096a 0x0096a RWE 0x2
  LOAD           0x0037b8 0x0000c4b5 0x000864b5 0x0013c 0x0013c R E 0x1
  LOAD           0x0038f4 0x0008b000 0x0008b000 0x001a0 0x001a0 R   0x2
  LOAD           0x003a94 0x00009000 0x00098000 0x00296 0x00296 R E 0x1
  LOAD           0x003d2c 0x00000400 0x00000400 0x00000 0x00800 RW  0x2
  LOAD           0x003d2c 0x00008000 0x00008000 0x00000 0x012c6 RW  0x2
  LOAD           0x003d2c 0x00008964 0x00008964 0x00000 0x00094 RW  0x2

 Section to Segment mapping:
  Segment Sections...
   00     codestart 
   01     .cinit 
   02     .text.1 .text.2 
   03     .TI.ramfunc 
   04     ramfuncs 
   05     FPUmathTables 
   06     .const_cla isrcodefuncs dclfuncs Cla1Prog 
   07     .stack 
   08     .data .bss .const_cla dclfuncs Cla1Prog 
   09     .data 
```

This is the output from TI's `ofd2000` tool for sections:
```
% ofd2000 --obj_display=none,sections test/testfiles_for_readelf/ti_c2000.elf 

OBJECT FILE:  test/testfiles_for_readelf/ti_c2000.elf

 Section Information

    id name                      load addr  run addr     size align alloc
    -- ----                      ---------  --------     ---- ----- -----
     0 (no name)                 0x00000000 0x00000000      0     0   N  
     1 codestart                 0x00080000 0x00080000    0x2     4   Y  
     2 .cinit                    0x00080004 0x00080004   0x3a     4   Y  
     3 .init_array               0x00000000 0x00000000      0     4   Y  
     4 .switch                   0x00000000 0x00000000      0     4   N  
     5 .const                    0x00000000 0x00000000      0     4   N  
     6 .reset                    0x003fffc0 0x003fffc0      0     1   N  
     7 .stack                    0x00000400 0x00000400  0x400     2   Y  
     8 .data                     0x00008964 0x00008964   0x4a     2   Y  
     9 .bss                      0x00008000 0x00008000  0x963     2   Y  
    10 .sysmem                   0x00000000 0x00000000      0     1   N  
    11 .TI.ramfunc               0x00086000 0x0000c000  0x4b5     2   Y  
    12 ramfuncs                  0x000864b5 0x0000c4b5   0x9e     1   Y  
    13 SFRA_F32_Data             0x00000000 0x00000000      0    64   N  
    14 FPUmathTables             0x0008b000 0x0008b000   0xd0     2   Y  
    15 .scratchpad               0x00000000 0x00000000      0     1   N  
    16 .bss_cla                  0x00000000 0x00000000      0     1   Y  
    17 controlVariables          0x00000000 0x00000000      0     1   N  
    18 .const_cla                0x00009000 0x00009000      0     1   Y  
    19 isrcodefuncs              0x00098000 0x00009000  0x14b     1   Y  
    20 dclfuncs                  0x0000914b 0x0000914b      0     1   Y  
    21 Cla1Prog                  0x00009000 0x00009000      0     4   Y  
    22 .ppdata                   0x00000000 0x00000000      0     1   N  
    23 .debug_types              0x00000000 0x00000000 0xa4a0     1   N  
    24 .debug_info               0x00000000 0x00000000 0x3db2     1   N  
    25 .debug_line               0x00000000 0x00000000 0x5cd6     1   N  
    26 .debug_frame              0x00000000 0x00000000  0xdb0     4   N  
    27 .text.1                   0x00081000 0x00081000 0x1000     1   Y  
    28 .text.2                   0x00082000 0x00082000  0x6cf     1   Y  
    29 .debug_abbrev             0x00000000 0x00000000  0x48a     0   N  
    30 .debug_str                0x00000000 0x00000000 0x1cf3     0   N  
    31 .debug_aranges            0x00000000 0x00000000  0xb38     0   N  
    32 .debug_pubnames           0x00000000 0x00000000 0x19fd     0   N  
    33 __TI_build_attributes     0x00000000 0x00000000   0x3a     0   N  
    34 .symtab                   0x00000000 0x00000000 0x4040     0   N  
    35 .TI.section.flags         0x00000000 0x00000000   0x37     0   N  
    36 .strtab                   0x00000000 0x00000000 0x21b7     0   N  
    37 .shstrtab                 0x00000000 0x00000000  0x193     0   N  
```
and for segments:
```
% ofd2000 --obj_display=none,segments test/testfiles_for_readelf/ti_c2000.elf 

OBJECT FILE:  test/testfiles_for_readelf/ti_c2000.elf

 Program Segment Table

    id type    load addr  run addr   file size mem size flags align
    -- ----    ---------  --------   --------- -------- ----- -----
    0  PT_LOAD 0x00080000 0x00080000 0x2       0x2      r-x   4    
    1  PT_LOAD 0x00080004 0x00080004 0x3a      0x3a     r--   4    
    2  PT_LOAD 0x00081000 0x00081000 0x16cf    0x16cf   r-x   1    
    3  PT_LOAD 0x00086000 0x0000c000 0x4b5     0x4b5    rwx   2    
    4  PT_LOAD 0x000864b5 0x0000c4b5 0x9e      0x9e     r-x   1    
    5  PT_LOAD 0x0008b000 0x0008b000 0xd0      0xd0     r--   2    
    6  PT_LOAD 0x00098000 0x00009000 0x14b     0x14b    r-x   1    
    7  PT_LOAD 0x00000400 0x00000400 0         0x400    rw-   2    
    8  PT_LOAD 0x00008000 0x00008000 0         0x963    rw-   2    
    9  PT_LOAD 0x00008964 0x00008964 0         0x4a     rw-   2    
```

Now compare the output from `readelf.py` with my modifications:
```
% python3.11 scripts/readelf.py -T -S test/testfiles_for_readelf/ti_c2000.elf 
There are 38 section headers, starting at offset 0x235f8

Section Headers:
  [Nr] Name              Type            LoadAddr RunAddr  Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 00000000 000000 000000 00      0   0  0
  [ 1] codestart         PROGBITS        00080000 00080000 000034 000002 00  AX  0   0  4
  [ 2] .cinit            PROGBITS        00080004 00080004 000038 00003a 00   A  0   0  4
  [ 3] .init_array       NOBITS          00000000 00000000 003d30 000000 00  WA  0   0  4
  [ 4] .switch           NOBITS          00000000 00000000 000000 000000 00      0   0  4
  [ 5] .const            NOBITS          00000000 00000000 000000 000000 00      0   0  4
  [ 6] .reset            NOBITS          003fffc0 003fffc0 000000 000000 00      0   0  1
  [ 7] .stack            NOBITS          00000400 00000400 003d2c 000400 00  WA  0   0  2
  [ 8] .data             NOBITS          00008964 00008964 003d2c 00004a 00  WA  0   0  2
  [ 9] .bss              NOBITS          00008000 00008000 003d2c 000963 00  WA  0   0  2
  [10] .sysmem           NOBITS          00000000 00000000 000000 000000 00      0   0  1
  [11] .TI.ramfunc       PROGBITS        00086000 0000c000 002e4c 0004b5 00 WAX  0   0  2
  [12] ramfuncs          PROGBITS        000864b5 0000c4b5 0037b8 00009e 00  AX  0   0  1
  [13] SFRA_F32_Data     NOBITS          00000000 00000000 000000 000000 00      0   0 64
  [14] FPUmathTables     PROGBITS        0008b000 0008b000 0038f4 0000d0 00   A  0   0  2
  [15] .scratchpad       NOBITS          00000000 00000000 000000 000000 00      0   0  1
  [16] .bss_cla          NOBITS          00000000 00000000 003d30 000000 00  WA  0   0  1
  [17] controlVariables  NOBITS          00000000 00000000 000000 000000 00      0   0  1
  [18] .const_cla        NOBITS          0009814e 00009000 003d30 000000 00   A  0   0  1
  [19] isrcodefuncs      PROGBITS        00098000 00009000 003a94 00014b 00  AX  0   0  1
  [20] dclfuncs          NOBITS          0009814e 0000914b 003d30 000000 00   A  0   0  1
  [21] Cla1Prog          NOBITS          0009814e 00009000 003d30 000000 00   A  0   0  4
  [22] .ppdata           NOBITS          00000000 00000000 000000 000000 00      0   0  1
  [23] .debug_types      PROGBITS        00000000 00000000 003d30 00a4a0 00      0   0  1
  [24] .debug_info       PROGBITS        00000000 00000000 00e1d0 003db2 00      0   0  1
  [25] .debug_line       PROGBITS        00000000 00000000 011f82 005cd6 00      0   0  1
  [26] .debug_frame      PROGBITS        00000000 00000000 017c58 000db0 00      0   0  4
  [27] .text.1           PROGBITS        00081000 00081000 0000ac 001000 00  AX  0   0  1
  [28] .text.2           PROGBITS        00082000 00082000 0020ac 0006cf 00  AX  0   0  1
  [29] .debug_abbrev     PROGBITS        00000000 00000000 018a08 00048a 00      0   0  0
  [30] .debug_str        PROGBITS        00000000 00000000 018e92 001cf3 00      0   0  0
  [31] .debug_aranges    PROGBITS        00000000 00000000 01ab85 000b38 00      0   0  0
  [32] .debug_pubnames   PROGBITS        00000000 00000000 01b6bd 0019fd 00      0   0  0
  [33] __TI_build_attrib <unknown>       00000000 00000000 01d0ba 00003a 00      0   0  0
  [34] .symtab           SYMTAB          00000000 00000000 01d0f4 004040 10     36 735  0
  [35] .TI.section.flags <unknown>       00000000 00000000 021134 000037 00      0   0  0
  [36] .strtab           STRTAB          00000000 00000000 02116b 0021b7 01   S  0   0  0
  [37] .shstrtab         STRTAB          00000000 00000000 023322 000193 01   S  0   0  0
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  p (processor specific)
```
and
```
% python3.11 scripts/readelf.py -T -l test/testfiles_for_readelf/ti_c2000.elf 

Elf file type is EXEC (Executable file)
Entry point is 0x80000
There are 10 program headers, starting at offset 144568

Program Headers:
  Segment Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
   00     LOAD           0x000034 0x00080000 0x00080000 0x00002 0x00002 R E 0x4
   01     LOAD           0x000038 0x00080004 0x00080004 0x0003a 0x0003a R   0x4
   02     LOAD           0x0000ac 0x00081000 0x00081000 0x016cf 0x016cf R E 0x1
   03     LOAD           0x002e4c 0x0000c000 0x00086000 0x004b5 0x004b5 RWE 0x2
   04     LOAD           0x0037b8 0x0000c4b5 0x000864b5 0x0009e 0x0009e R E 0x1
   05     LOAD           0x0038f4 0x0008b000 0x0008b000 0x000d0 0x000d0 R   0x2
   06     LOAD           0x003a94 0x00009000 0x00098000 0x0014b 0x0014b R E 0x1
   07     LOAD           0x003d2c 0x00000400 0x00000400 0x00000 0x00400 RW  0x2
   08     LOAD           0x003d2c 0x00008000 0x00008000 0x00000 0x00963 RW  0x2
   09     LOAD           0x003d2c 0x00008964 0x00008964 0x00000 0x0004a RW  0x2

 Section to Segment mapping:
  Segment Sections...
   00     codestart 
   01     .cinit 
   02     .text.1 .text.2 
   03     .TI.ramfunc 
   04     ramfuncs 
   05     FPUmathTables 
   06     .const_cla isrcodefuncs dclfuncs Cla1Prog 
   07     .stack 
   08     .data .bss .const_cla dclfuncs Cla1Prog 
   09     .data 
```

Many thanks for all your time and effort in maintaining this amazing library.
All the best,
Dimitri

eliben commented:
Thanks for the context.

Adding features / enums / etc. to make pyelftools understand more object files emitted by TI compilers - ✅ 
Changing readelf.py to match the output of some TI's proprietary toolchain - ❌ 

The only reason readelf.py exists is for high-level testing compared to GNU readelf.
